examples/diff.c

```diff
@@ -86,49 +151,80 @@ int lg2_diff(git_repository *repo, int argc, char *argv[])
 	 *  * &lt;sha1&gt;
 	 *  * --cached
 	 *  * --nocache (don't use index data in diff at all)
+	 *  * --no-index &lt;file1&gt; &lt;file2&gt;
 	 *  * nothing
 	 *
 	 * Currently ranged arguments like &lt;sha1&gt;..&lt;sha2&gt; and &lt;sha1&gt;...&lt;sha2&gt;
 	 * are not supported in this example
 	 */
 
-	if (o.treeish1)
-		treeish_to_tree(&t1, repo, o.treeish1);
-	if (o.treeish2)
-		treeish_to_tree(&t2, repo, o.treeish2);
+	if (o.no_index >= 0) {
+		if (o.treeish1 && o.treeish2) {
+			file1_str = read_file(o.treeish1);
+		   if (file1_str == NULL) {
+		      usage("file cannot be readed", o.treeish1);
+		   }
+		  	file2_str = read_file(o.treeish2);
+		   if (file2_str == NULL) {
+		      usage("file cannot be readed", o.treeish2);
+		   }
+		} else {
+		   usage("two files should be provided as arguments", NULL);
+		}
+	}
 
-	if (t1 && t2)
+	if (file1_str && file2_str) {
+	   git_buf buf = GIT_BUF_INIT;
 		check_lg2(
-			git_diff_tree_to_tree(&diff, repo, t1, t2, &o.diffopts),
-			"diff trees", NULL);
-	else if (o.cache != CACHE_NORMAL) {
-		if (!t1)
-			treeish_to_tree(&t1, repo, "HEAD");
-
-		if (o.cache == CACHE_NONE)
+			git_patch_from_buffers(&patch, file1_str, strlen(file1_str), o.treeish1, file2_str, strlen(file2_str), o.treeish2, &o.diffopts),
+			"patch buffers", NULL);
+		check_lg2(
+		   git_patch_to_buf(&buf, patch),
```

Comment:
Though not your fault, I kinda dislike that style…

This is a styleguide question for maintainers, do we forgo `check_lg2` if that's too heavy-handed/confusing, or switch to one of our `if ((err = func()) < 0 || …)` patterns instead, if it's mostly related to one task (here loading a diff from a buffer) ?

Reply:
Just this review remains, not sure how to fix this: 

Is it something like this?
```
check_lg2(
    !git_patch_from_buffers(&patch, file1_str, strlen(file1_str), o.treeish1,
        file2_str, strlen(file2_str), o.treeish2, &o.diffopts) &&
    !git_patch_to_buf(&buf, patch) &&
    !git_diff_from_buffer(&diff, buf.ptr, buf.size),
    "diff buffers", NULL);
```

Reply:
@pks-t Any opinions ? I'm not a fan of having `check_lg2` in example code in the first place (it makes the error handling brittle), and now it impacts readability (somewhat).

I don't think its worth blocking that PR on that though, that's just food for thought for the next time I look at example code.